core/cache/resident.go

```diff
@@ -75,16 +87,80 @@ func (m *residentManager) new() *Resident {
 		nextResourceId: func() uint64 { return m.resourceCount.next() },
 		workers:        make(map[uint64]worker.Worker),
 	}
+
+	m.mu.Lock()
 	m.residents[r.id] = r
+	m.mu.Unlock()
+
 	return r
 }
 
-func (m *residentManager) evict(id uint64) {
-	// TODO (manadart 2019-04-17): TBC when the mark/sweep logic is added.
+// mark sets all of the manager's residents to be stale.
+// The boolean return indicates if there were any residents to mark.
+func (m *residentManager) mark() bool {
+	m.mu.Lock()
+	for _, r := range m.residents {
+		r.setStale(true)
+	}
+	m.mu.Unlock()
+
+	return len(m.residents) > 0
+}
+
+// sweep removes stale cache residents in descending order of ID.
+func (m *residentManager) sweep() <-chan struct{} {
+	removalIds, removalMessages := m.evictions()
+
+	finished := make(chan struct{})
+
+	go func() {
+	loop:
+		for _, id := range removalIds {
+			select {
+			case m.removals <- removalMessages[id]:
+			case <-m.dying:
+				logger.Debugf("aborting cache sweep")
+				break loop
+			}
+		}
+		close(finished)
+	}()
+
+	return finished
+}
+
+// evictions iterates over the cache residents and generates a map of
+// eviction messages and a slice for determining order of eviction.
+// Because the IDs are supplied in increasing order,
+// this ensures we never remove a resident's model before the resident itself.
+func (m *residentManager) evictions() ([]uint64, map[uint64]interface{}) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Create a descending order slice of stale IDs,
+	// and a map of resident removal messages.
+	var removalIds []uint64
+	removalMessages := make(map[uint64]interface{})
+	for id, r := range m.residents {
+		if r.isStale() {
+			if r.removalMessage == nil {
+				logger.Warningf("cache resident %d has no removal message; skipping eviction", id)
+				continue
+			}
+
+			removalIds = append(removalIds, id)
+			removalMessages[id] = r.removalMessage
```

Comment:
In theory you could end up with more `removalIds` than `removalMessages`, because id could be featured twice with two different removal messages and the last removalMessage would win. We should probably comment this is the case, although in reality that won't matter, but for future selves we should state that's fine.

Reply:
I don't think you can. The removal IDs are generated from the keys of the resident map. They will be unique.

In any case, you are right - the removal logic is such that if the typed entity is not found, removal is a no-op.